# Для тренировки модели нужно установить turicreate
Linux or mac
python 3.6 - 3.8

`pip install turicreate`

### На вход:
- _attend_features.csv (с нашими фичами)
- _user_features.csv (с нашими фичами)
- _groups.csv

### Выход:
- predicts/predict_v01.csv


In [2]:
import turicreate as tc
import pandas as pd

In [3]:
actions = tc.SFrame.read_csv('_attend_features.csv')

Finished parsing file /media/vasiliev/SSD_Data/_dev/ezee_may_2003/ds/_attend_features.csv

Parsing completed. Parsed 100 lines in 0.302705 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[int,int,int,float,str,str,float,str,str,str,str,float,float]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 219940 lines. Lines per second: 651134

Finished parsing file /media/vasiliev/SSD_Data/_dev/ezee_may_2003/ds/_attend_features.csv

Parsing completed. Parsed 600390 lines in 0.59433 secs.

In [4]:
#items = tc.SFrame.read_csv('groups.csv')

There are a variety of machine learning techniques that can be used to build a recommender model. Turi Create provides a method turicreate.recommender.create that will automatically choose an appropriate model for your data set. More details on choosing a model: https://apple.github.io/turicreate/docs/userguide/recommender/choosing-a-model.html

First we create a random split of the data to produce a validation set that can be used to evaluate the model.

### Чему учим модель

Мы учим модель предсказывать логарифм от количества посещений юзером курса (логарифм потому что разброс посещений курса пользователем высокий - от 1 до 64)

### Создание и обучение дефолтной модели

In [8]:
training_data, validation_data = tc.recommender.util.random_split_by_user(actions, 'user_id', 'group_id')
model = tc.recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits')

Preparing data set.

Data has 597961 observations with 50881 users and 24534 items.

Data prepared in: 0.39s

Training ranking_factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 32       |

| regularization                 | L2 Regularization on Factors                     | 1e-09    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-09    |

| ranking_regularization         | Rank-based Regularization Weight                 | 0.25     |

| max_iterations                 | Maximum Number of Iterations                     | 25       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 74745 / 597961 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 4.16667           | Not Viable                               |

| 1       | 1.04167           | Not Viable                               |

| 2       | 0.260417          | Not Viable                               |

| 3       | 0.0651042         | Not Viable                               |

| 4       | 0.016276          | 1.33577                                  |

| 5       | 0.00813802        | 1.81101                                  |

| 6       | 0.00406901        | 2.08359                                  |

| 7       | 0.00203451        | 2.16795                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.016276          | 1.33577                                  |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 52us         | 2.77823           | 1.19033               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 468.172ms    | 2.03111           | 1.08959               | 0.016276    |

| 2       | 921.405ms    | 1.58929           | 0.976142              | 0.016276    |

| 3       | 1.36s        | 1.36365           | 0.917098              | 0.016276    |

| 4       | 1.79s        | 1.22625           | 0.877196              | 0.016276    |

| 5       | 2.23s        | 1.13007           | 0.848961              | 0.016276    |

| 10      | 4.25s        | 0.877373          | 0.768772              | 0.016276    |

| 20      | 8.12s        | 0.694482          | 0.703868              | 0.016276    |

| 25      | 10.00s       | 0.646922          | 0.685629              | 0.016276    |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.624801

Final training RMSE: 0.671059

In [18]:
# параметры модели
model

Class                            : RankingFactorizationRecommender

Schema
------
User ID                          : user_id
Item ID                          : group_id
Target                           : log_visits
Additional observation features  : 10
User side features               : []
Item side features               : []

Statistics
----------
Number of observations           : 597961
Number of users                  : 50881
Number of items                  : 24534

Training summary
----------------
Training time                    : 12.0792

Model Parameters
----------------
Model class                      : RankingFactorizationRecommender
num_factors                      : 32
binary_target                    : 0
side_data_factorization          : 1
solver                           : auto
nmf                              : 0
max_iterations                   : 25

Regularization Settings
-----------------------
regularization                   : 0.0
regularization_type          

### Получение предсказаний от дефолтной модели

In [11]:
default_recommendations = model.recommend()

recommendations finished on 1000/50881 queries. users per second: 5323.54

recommendations finished on 2000/50881 queries. users per second: 5384.51

recommendations finished on 3000/50881 queries. users per second: 5377.35

recommendations finished on 4000/50881 queries. users per second: 5374

recommendations finished on 5000/50881 queries. users per second: 5389.5

recommendations finished on 6000/50881 queries. users per second: 5393.54

recommendations finished on 7000/50881 queries. users per second: 5378.99

recommendations finished on 8000/50881 queries. users per second: 5373.5

recommendations finished on 9000/50881 queries. users per second: 5371.4

recommendations finished on 10000/50881 queries. users per second: 5368.27

recommendations finished on 11000/50881 queries. users per second: 5361.7

recommendations finished on 12000/50881 queries. users per second: 5354.5

recommendations finished on 13000/50881 queries. users per second: 5359.22

recommendations finished on 14000/50881 queries. users per second: 5359.39

recommendations finished on 15000/50881 queries. users per second: 5365.62

recommendations finished on 16000/50881 queries. users per second: 5366.85

recommendations finished on 17000/50881 queries. users per second: 5370.08

recommendations finished on 18000/50881 queries. users per second: 5372.43

recommendations finished on 19000/50881 queries. users per second: 5374.98

recommendations finished on 20000/50881 queries. users per second: 5376.89

recommendations finished on 21000/50881 queries. users per second: 5380.51

recommendations finished on 22000/50881 queries. users per second: 5382.99

recommendations finished on 23000/50881 queries. users per second: 5381.84

recommendations finished on 24000/50881 queries. users per second: 5378.63

recommendations finished on 25000/50881 queries. users per second: 5378.05

recommendations finished on 26000/50881 queries. users per second: 5379.85

recommendations finished on 27000/50881 queries. users per second: 5382.02

recommendations finished on 28000/50881 queries. users per second: 5383.37

recommendations finished on 29000/50881 queries. users per second: 5383.75

recommendations finished on 30000/50881 queries. users per second: 5385.52

recommendations finished on 31000/50881 queries. users per second: 5387.13

recommendations finished on 32000/50881 queries. users per second: 5388.47

recommendations finished on 33000/50881 queries. users per second: 5389.65

recommendations finished on 34000/50881 queries. users per second: 5389.19

recommendations finished on 35000/50881 queries. users per second: 5389.63

recommendations finished on 36000/50881 queries. users per second: 5388.33

recommendations finished on 37000/50881 queries. users per second: 5386.53

recommendations finished on 38000/50881 queries. users per second: 5378.07

recommendations finished on 39000/50881 queries. users per second: 5360.2

recommendations finished on 40000/50881 queries. users per second: 5331.01

recommendations finished on 41000/50881 queries. users per second: 5282.08

recommendations finished on 42000/50881 queries. users per second: 5220.04

recommendations finished on 43000/50881 queries. users per second: 5161.05

recommendations finished on 44000/50881 queries. users per second: 5107.87

recommendations finished on 45000/50881 queries. users per second: 5045.64

recommendations finished on 46000/50881 queries. users per second: 4981.29

recommendations finished on 47000/50881 queries. users per second: 4918.78

recommendations finished on 48000/50881 queries. users per second: 4856.43

recommendations finished on 49000/50881 queries. users per second: 4798.89

recommendations finished on 50000/50881 queries. users per second: 4682.73

### Посмотрим рекоммендации для отдельных юзеров

In [14]:
df_groups = pd.read_csv('groups.csv')

def display_history_and_predictions_for_user(user_id, train_dataset, predictions):
    print('----------------------------------------------------------------------------')
    print('User', user_id, 'history:')
    display(train_dataset[train_dataset['user_id'] == user_id][
        'group_id','is_online','session','направление 1','направление 2'])
    predictions = predictions[predictions['user_id'] == user_id].to_dataframe()
    print('User recommendations:')
    predicted_groups = predictions['group_id'].to_list()
    display(df_groups[df_groups['уникальный номер'].isin(predicted_groups)].drop(columns=[
        'адрес площадки','расписание в активных периодах','расписание в закрытых периодах','расписание в плановом периоде']))

In [16]:
user_id = 101352882

display_history_and_predictions_for_user(user_id, actions, results)

----------------------------------------------------------------------------
User 101352882 history:


Columns:
	group_id	int
	is_online	str
	session	str
	направление 1	str
	направление 2	str

Rows: Unknown

Data:
+-----------+-----------+---------+-----------------------+
|  group_id | is_online | session |     направление 1     |
+-----------+-----------+---------+-----------------------+
| 801348234 |   online  |  mornig | Физическая активность |
| 801348838 |  offline  |  mornig | Физическая активность |
| 801348944 |  offline  |  mornig |         Танцы         |
| 801349048 |  offline  |  mornig | Физическая активность |
| 801350227 |   online  |  mornig |       Творчество      |
| 801350380 |   online  |  mornig |         Танцы         |
| 801351413 |   online  |  mornig |       Рисование       |
| 801358464 |   online  |  mornig | Физическая активность |
| 801358465 |   online  |  mornig |         Танцы         |
| 801358466 |   online  |  mornig | Физическая активность |
+-----------+-----------+---------+-----------------------+
+-------------------------------+
|         направление 2         |
+-------------------------------+
|       ОНЛАЙН Гимнастика       |
|           Гимнастика          |
|             Танцы             |
|              ОФП              |
| ОНЛАЙН Художественно-прикл... |
|          ОНЛАЙН Танцы         |
|        ОНЛАЙН Рисование       |
|       ОНЛАЙН Гимнастика       |
|          ОНЛАЙН Танцы         |
|       ОНЛАЙН Гимнастика       |
+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

User recommendations:


уникальный номер          направление 1     направление 2   
882          801348532              Рисование  ОНЛАЙН Рисование  \
1708         801347075  Физическая активность        ОНЛАЙН ОФП   
1717         801346764  Физическая активность        ОНЛАЙН ОФП   
1730         801348933  Физическая активность        ОНЛАЙН ОФП   
1739         801349699  Физическая активность        ОНЛАЙН ОФП   
1765         801350565  Физическая активность        ОНЛАЙН ОФП   
1769         801348457  Физическая активность        ОНЛАЙН ОФП   
3915         801348042              Рисование  ОНЛАЙН Рисование   
3971         801350767              Рисование  ОНЛАЙН Рисование   
7460         801348374                  Танцы             Танцы   

                                    направление 3   
882                   ОНЛАЙН Акварельная живопись  \
1708  ОНЛАЙН Физкультурно-оздоровительные занятия   
1717  ОНЛАЙН Физкультурно-оздоровительные занятия   
1730  ОНЛАЙН Физкультурно-оздоровительные занятия   
1739  ОНЛАЙН Физкультурно-оздоровительные занятия   
1765  ОНЛАЙН Физкультурно-оздоровительные занятия   
1769  ОНЛАЙН Физкультурно-оздоровительные занятия   
3915                     ОНЛАЙН Масляная живопись   
3971           ОНЛАЙН Различные техники рисования   
7460                               Танцы для всех   

                              округ площадки   
882    Зеленоградский административный округ  \
1708  Северо-Западный административный округ   
1717        Восточный административный округ   
1730         Западный административный округ   
1739            Южный административный округ   
1765        Восточный административный округ   
1769       Северный административный округ,    
3915   Зеленоградский административный округ   
3971                                     NaN   
7460      Центральный административный округ   

                              район площадки  
882              муниципальный округ Крюково  
1708              муниципальный округ Щукино  
1717  муниципальный округ Северное Измайлово  
1730           муниципальный округ Можайский  
1739         муниципальный округ Даниловский  
1765           муниципальный округ Гольяново  
1769             муниципальный округ Сокол,   
3915             муниципальный округ Крюково  
3971                                     NaN  
7460            муниципальный округ Тверской

In [17]:
user_id = 101354253

display_history_and_predictions_for_user(user_id, actions, results)

----------------------------------------------------------------------------
User 101354253 history:


Columns:
	group_id	int
	is_online	str
	session	str
	направление 1	str
	направление 2	str

Rows: Unknown

Data:
+-----------+-----------+-----------+-----------------------+-------------------+
|  group_id | is_online |  session  |     направление 1     |   направление 2   |
+-----------+-----------+-----------+-----------------------+-------------------+
| 801347802 |   online  |   mornig  | Физическая активность |     ОНЛАЙН ОФП    |
| 801348667 |   online  |  evening  | Физическая активность | ОНЛАЙН Гимнастика |
| 801348794 |   online  |   mornig  | Физическая активность | ОНЛАЙН Гимнастика |
| 801349796 |  offline  |   mornig  | Физическая активность |     Гимнастика    |
| 801349875 |  offline  | aftennoon | Физическая активность |     Гимнастика    |
| 801350556 |   online  |   mornig  | Физическая активность |     ОНЛАЙН ОФП    |
| 801350568 |   online  |   mornig  | Физическая активность |     ОНЛАЙН ОФП    |
| 801361065 |  offline  |   mornig  |       Рисование       |     Рисование     |
+-----------+-----------+-----------+-----------------------+-------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

User recommendations:


уникальный номер          направление 1      направление 2   
1585          801352052  Физическая активность  ОНЛАЙН Гимнастика  \
2160          801351037  Физическая активность  ОНЛАЙН Гимнастика   
3075          801351857  Физическая активность  ОНЛАЙН Гимнастика   
3788          801347329  Физическая активность  ОНЛАЙН Гимнастика   
4545          801348670  Физическая активность  ОНЛАЙН Гимнастика   
4668          801348888  Физическая активность  ОНЛАЙН Гимнастика   
8154          801349632  Физическая активность  ОНЛАЙН Гимнастика   
12628         801346690  Физическая активность  ОНЛАЙН Гимнастика   
12646         801351151  Физическая активность  ОНЛАЙН Гимнастика   
20411         801365938  Физическая активность  ОНЛАЙН Гимнастика   

                                     направление 3   
1585                                   ОНЛАЙН Йога  \
2160   ОНЛАЙН Адаптивная и тонизирующая гимнастика   
3075                                   ОНЛАЙН Йога   
3788      ОНЛАЙН Корригирующая гимнастика для глаз   
4545                                   ОНЛАЙН Йога   
4668                                   ОНЛАЙН Йога   
8154                             ОНЛАЙН Гимнастика   
12628  ОНЛАЙН Адаптивная и тонизирующая гимнастика   
12646  ОНЛАЙН Адаптивная и тонизирующая гимнастика   
20411  ОНЛАЙН Адаптивная и тонизирующая гимнастика   

                             округ площадки   
1585     Центральный административный округ  \
2160                                    NaN   
3075                                    NaN   
3788       Восточный административный округ   
4545     Центральный административный округ   
4668           Южный административный округ   
8154           Южный административный округ   
12628  Юго-Восточный административный округ   
12646      Восточный административный округ   
20411      Восточный административный округ   

                               район площадки  
1585            муниципальный округ Таганский  
2160                                      NaN  
3075                                      NaN  
3788   муниципальный округ Северное Измайлово  
4545             муниципальный округ Тверской  
4668             муниципальный округ Нагорный  
8154             муниципальный округ Царицыно  
12628             муниципальный округ Люблино  
12646           муниципальный округ Восточный  
20411           муниципальный округ Гольяново

In [19]:
user_id = 101385865

display_history_and_predictions_for_user(user_id, actions, results)

----------------------------------------------------------------------------
User 101385865 history:


Columns:
	group_id	int
	is_online	str
	session	str
	направление 1	str
	направление 2	str

Rows: Unknown

Data:
+-----------+-----------+-----------+-----------------------+
|  group_id | is_online |  session  |     направление 1     |
+-----------+-----------+-----------+-----------------------+
| 801346923 |   online  |   mornig  |         Танцы         |
| 801348309 |   online  |   mornig  |       Творчество      |
| 801348946 |   online  |   mornig  | Физическая активность |
| 801349093 |   online  |   mornig  |      Образование      |
| 801350488 |  offline  | aftennoon | Физическая активность |
| 801350739 |  offline  |   mornig  |         Танцы         |
| 801350759 |  offline  |   mornig  |         Танцы         |
| 801351106 |   online  |   mornig  |       Творчество      |
| 801351519 |  offline  |   mornig  |         Танцы         |
| 801351888 |  offline  |   mornig  |         Танцы         |
+-----------+-----------+-----------+-----------------------+
+--------------------------+
|      направление 2       |
+--------------------------+
|       ОНЛАЙН Танцы       |
|  ОНЛАЙН Красота и стиль  |
|    ОНЛАЙН Гимнастика     |
| ОНЛАЙН Иностранные языки |
|    Фитнес, тренажеры     |
|          Танцы           |
|          Танцы           |
|  ОНЛАЙН Красота и стиль  |
|          Танцы           |
|          Танцы           |
+--------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

User recommendations:


уникальный номер          направление 1   
1715         801349571  Физическая активность  \
1717         801346764  Физическая активность   
1730         801348933  Физическая активность   
1738         801347634  Физическая активность   
1739         801349699  Физическая активность   
1749         801351282  Физическая активность   
1750         801346908  Физическая активность   
1762         801351112  Физическая активность   
6167         801349321             Творчество   
6640         801348492             Творчество   

                            направление 2   
1715                           ОНЛАЙН ОФП  \
1717                           ОНЛАЙН ОФП   
1730                           ОНЛАЙН ОФП   
1738                           ОНЛАЙН ОФП   
1739                           ОНЛАЙН ОФП   
1749                           ОНЛАЙН ОФП   
1750                           ОНЛАЙН ОФП   
1762                           ОНЛАЙН ОФП   
6167  Художественно-прикладное творчество   
6640  Художественно-прикладное творчество   

                                    направление 3   
1715  ОНЛАЙН Физкультурно-оздоровительные занятия  \
1717  ОНЛАЙН Физкультурно-оздоровительные занятия   
1730  ОНЛАЙН Физкультурно-оздоровительные занятия   
1738  ОНЛАЙН Физкультурно-оздоровительные занятия   
1739  ОНЛАЙН Физкультурно-оздоровительные занятия   
1749  ОНЛАЙН Физкультурно-оздоровительные занятия   
1750  ОНЛАЙН Физкультурно-оздоровительные занятия   
1762  ОНЛАЙН Физкультурно-оздоровительные занятия   
6167                       Рукоделие и творчество   
6640           Текстильный дизайн, кройка и шитье   

                              округ площадки   
1715  Северо-Западный административный округ  \
1717        Восточный административный округ   
1730         Западный административный округ   
1738     Юго-Западный административный округ   
1739            Южный административный округ   
1749            Южный административный округ   
1750      Восточный административный округ,    
1762         Западный административный округ   
6167     Юго-Западный административный округ   
6640     Юго-Западный административный округ   

                              район площадки  
1715            муниципальный округ Строгино  
1717  муниципальный округ Северное Измайлово  
1730           муниципальный округ Можайский  
1738              муниципальный округ Зюзино  
1739         муниципальный округ Даниловский  
1749  муниципальный округ Бирюлево Восточное  
1750         муниципальный округ Измайлово,   
1762    муниципальный округ Ново-Переделкино  
6167        муниципальный округ Южное Бутово  
6640        муниципальный округ Южное Бутово

### Делаем сравнение различных моделей

Описание всех моделей, заложенных в turicreate, доступно по ссылке:

https://apple.github.io/turicreate/docs/api/turicreate.toolkits.recommender.html

In [20]:
model_0 = tc.recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits')
model_1 = tc.item_similarity_recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits')
model_2 = tc.popularity_recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits')
model_3 = tc.factorization_recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits')
model_4 = tc.ranking_factorization_recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits')

Preparing data set.

Data has 597961 observations with 50881 users and 24534 items.

Data prepared in: 0.383683s

Training ranking_factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 32       |

| regularization                 | L2 Regularization on Factors                     | 1e-09    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-09    |

| ranking_regularization         | Rank-based Regularization Weight                 | 0.25     |

| max_iterations                 | Maximum Number of Iterations                     | 25       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 74745 / 597961 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 4.16667           | Not Viable                               |

| 1       | 1.04167           | Not Viable                               |

| 2       | 0.260417          | 0.438846                                 |

| 3       | 0.130208          | 0.924721                                 |

+---------+-------------------+------------------------------------------+

| Final   | 0.260417          | 0.438846                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 58us         | 2.77822           | 1.19034               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 473.394ms    | 3.36485           | 1.48941               | 0.260417    |

| 2       | 916.802ms    | 1.73916           | 1.0496                | 0.260417    |

| 3       | 1.33s        | 0.880088          | 0.732584              | 0.260417    |

| 4       | 1.73s        | 0.573706          | 0.607905              | 0.260417    |

| 5       | 2.10s        | 0.419765          | 0.533319              | 0.260417    |

| 10      | 3.86s        | 0.185564          | 0.374364              | 0.260417    |

| 20      | 7.12s        | 0.103839          | 0.285795              | 0.260417    |

| 25      | 8.68s        | 0.0904662         | 0.268129              | 0.260417    |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.0785142

Final training RMSE: 0.245509

Warning: Ignoring columns X1, направление 3, is_online, lesson_length, session, направление 1, направление 2, user_address, lat, long;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 597961 observations with 50881 users and 24534 items.

Data prepared in: 0.162217s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 6.236ms                        | 1.75       |

| 10.231ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 366.098ms                           | 0                | 0               |

| 793.747ms                           | 100              | 24534           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 0.813125s

Warning: Ignoring columns X1, направление 3, is_online, lesson_length, session, направление 1, направление 2, user_address, lat, long;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 597961 observations with 50881 users and 24534 items.

Data prepared in: 0.171603s

597961 observations to process; with 24534 unique items.

Preparing data set.

Data has 597961 observations with 50881 users and 24534 items.

Data prepared in: 0.389769s

Training factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 8        |

| regularization                 | L2 Regularization on Factors                     | 1e-08    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-10    |

| max_iterations                 | Maximum Number of Iterations                     | 50       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 74745 / 597961 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 4.16667           | Not Viable                               |

| 1       | 1.04167           | Not Viable                               |

| 2       | 0.260417          | 0.32967                                  |

| 3       | 0.130208          | 0.110857                                 |

| 4       | 0.0651042         | 0.258462                                 |

+---------+-------------------+------------------------------------------+

| Final   | 0.130208          | 0.110857                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 49us         | 1.4169            | 1.19034               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 122.224ms    | 1.02391           | 1.01178               | 0.130208    |

| 2       | 239.007ms    | 0.705449          | 0.839769              | 0.130208    |

| 3       | 356.528ms    | 0.599024          | 0.773789              | 0.130208    |

| 4       | 476.651ms    | 0.536072          | 0.731959              | 0.130208    |

| 5       | 595.326ms    | 0.490956          | 0.700439              | 0.130208    |

| 10      | 1.18s        | 0.38432           | 0.61957               | 0.130208    |

| 50      | 5.93s        | 0.253913          | 0.503192              | 0.130208    |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.243682

Final training RMSE: 0.492921

Preparing data set.

Data has 597961 observations with 50881 users and 24534 items.

Data prepared in: 0.381521s

Training ranking_factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 32       |

| regularization                 | L2 Regularization on Factors                     | 1e-09    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-09    |

| ranking_regularization         | Rank-based Regularization Weight                 | 0.25     |

| max_iterations                 | Maximum Number of Iterations                     | 25       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 74745 / 597961 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 4.16667           | Not Viable                               |

| 1       | 1.04167           | Not Viable                               |

| 2       | 0.260417          | Not Viable                               |

| 3       | 0.0651042         | 0.955653                                 |

| 4       | 0.0325521         | 0.677587                                 |

| 5       | 0.016276          | 1.60338                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.0325521         | 0.677587                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 66us         | 2.77824           | 1.19034               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 479.763ms    | 1.96562           | 1.08536               | 0.0325521   |

| 2       | 926.214ms    | 1.3364            | 0.921407              | 0.0325521   |

| 3       | 1.33s        | 1.05115           | 0.8371                | 0.0325521   |

| 4       | 1.72s        | 0.895316          | 0.78571               | 0.0325521   |

| 5       | 2.10s        | 0.797858          | 0.751181              | 0.0325521   |

| 10      | 3.92s        | 0.589865          | 0.666953              | 0.0325521   |

| 20      | 7.33s        | 0.467398          | 0.607238              | 0.0325521   |

| 25      | 8.97s        | 0.436898          | 0.590643              | 0.0325521   |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.408929

Final training RMSE: 0.567674

In [21]:
print(tc.recommender.util.compare_models(
    validation_data, 
    [model_0, model_1,model_2,model_3,model_4], 
    model_names=["Default model",
                 "Item similarity model",
                 'Popularity recommender',
                 'Factorization recommender',
                 'Ranking factorization recommender']))

PROGRESS: Evaluate model Default model

Precision and recall summary statistics by cutoff
+--------+-----------------------+------------------------+
| cutoff |     mean_precision    |      mean_recall       |
+--------+-----------------------+------------------------+
|   1    | 0.0013888888888888892 | 3.561253561253564e-05  |
|   2    | 0.0006944444444444446 | 3.5612535612535624e-05 |
|   3    | 0.0009259259259259257 | 0.00012820512820512826 |
|   4    |  0.001388888888888889 | 0.0002630240130240129  |
|   5    | 0.0013888888888888885 | 0.0003447233594292417  |
|   6    |  0.00162037037037037  |  0.000565181913221129  |
|   7    |  0.001388888888888889 |  0.000565181913221129  |
|   8    | 0.0012152777777777782 | 0.0005651819132211285  |
|   9    | 0.0010802469135802472 |  0.000565181913221129  |
|   10   | 0.0011111111111111111 |  0.001954070802110017  |
+--------+-----------------------+------------------------+
[10 rows x 3 columns]


Overall RMSE: 1.4839031983440507

Per User RMS

`Default model:             `Overall RMSE: 1.4839031983440507

`Item similarity model:     `Overall RMSE: 1.9387119289000498

`Popularity recommender:    `Overall RMSE: 0.9897593711404455

`Factorization recommender: `Overall RMSE: 1.0964792988739434

`Ranking factorization:     `Overall RMSE: 1.0114718154853628

### Берем результаты Ranking factorization модели

In [22]:
preds = model_4.recommend(k=10)
display(preds)

recommendations finished on 1000/50881 queries. users per second: 5354.35

recommendations finished on 2000/50881 queries. users per second: 5402.86

recommendations finished on 3000/50881 queries. users per second: 5446.88

recommendations finished on 4000/50881 queries. users per second: 5453.22

recommendations finished on 5000/50881 queries. users per second: 5462.93

recommendations finished on 6000/50881 queries. users per second: 5470.41

recommendations finished on 7000/50881 queries. users per second: 5475.9

recommendations finished on 8000/50881 queries. users per second: 5473.17

recommendations finished on 9000/50881 queries. users per second: 5470.58

recommendations finished on 10000/50881 queries. users per second: 5470.47

recommendations finished on 11000/50881 queries. users per second: 5473.63

recommendations finished on 12000/50881 queries. users per second: 5477.19

recommendations finished on 13000/50881 queries. users per second: 5478.79

recommendations finished on 14000/50881 queries. users per second: 5480.69

recommendations finished on 15000/50881 queries. users per second: 5483.23

recommendations finished on 16000/50881 queries. users per second: 5484.77

recommendations finished on 17000/50881 queries. users per second: 5480.6

recommendations finished on 18000/50881 queries. users per second: 5478.76

recommendations finished on 19000/50881 queries. users per second: 5473.82

recommendations finished on 20000/50881 queries. users per second: 5475.21

recommendations finished on 21000/50881 queries. users per second: 5475.24

recommendations finished on 22000/50881 queries. users per second: 5474.38

recommendations finished on 23000/50881 queries. users per second: 5476.2

recommendations finished on 24000/50881 queries. users per second: 5474.94

recommendations finished on 25000/50881 queries. users per second: 5475.07

recommendations finished on 26000/50881 queries. users per second: 5474.71

recommendations finished on 27000/50881 queries. users per second: 5475.31

recommendations finished on 28000/50881 queries. users per second: 5475.75

recommendations finished on 29000/50881 queries. users per second: 5475.52

recommendations finished on 30000/50881 queries. users per second: 5476.82

recommendations finished on 31000/50881 queries. users per second: 5477.04

recommendations finished on 32000/50881 queries. users per second: 5474.35

recommendations finished on 33000/50881 queries. users per second: 5471.8

recommendations finished on 34000/50881 queries. users per second: 5469.79

recommendations finished on 35000/50881 queries. users per second: 5470.78

recommendations finished on 36000/50881 queries. users per second: 5469.78

recommendations finished on 37000/50881 queries. users per second: 5471.12

recommendations finished on 38000/50881 queries. users per second: 5470.75

recommendations finished on 39000/50881 queries. users per second: 5454.99

recommendations finished on 40000/50881 queries. users per second: 5425.39

recommendations finished on 41000/50881 queries. users per second: 5391.88

recommendations finished on 42000/50881 queries. users per second: 5337.44

recommendations finished on 43000/50881 queries. users per second: 5262.46

recommendations finished on 44000/50881 queries. users per second: 5193.23

recommendations finished on 45000/50881 queries. users per second: 5128.9

recommendations finished on 46000/50881 queries. users per second: 5068.88

recommendations finished on 47000/50881 queries. users per second: 5013.83

recommendations finished on 48000/50881 queries. users per second: 4958.91

recommendations finished on 49000/50881 queries. users per second: 4904.74

recommendations finished on 50000/50881 queries. users per second: 4848.87

Columns:
	user_id	int
	group_id	int
	score	float
	rank	int

Rows: 508810

Data:
+-----------+-----------+--------------------+------+
|  user_id  |  group_id |       score        | rank |
+-----------+-----------+--------------------+------+
| 101370217 | 801362030 | 1.8778277140142543 |  1   |
| 101370217 | 801352480 | 1.8631233822586162 |  2   |
| 101370217 | 801354202 | 1.8266987528683765 |  3   |
| 101370217 | 801366419 |  1.80930315219184  |  4   |
| 101370217 | 801349419 | 1.7973920505525691 |  5   |
| 101370217 | 801348731 | 1.7943034868719203 |  6   |
| 101370217 | 801370008 | 1.7924303326787574 |  7   |
| 101370217 | 801373704 | 1.7881144460322482 |  8   |
| 101370217 | 801359318 | 1.7847448688032252 |  9   |
| 101370217 | 801350830 | 1.7810807701232059 |  10  |
+-----------+-----------+--------------------+------+
[508810 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### Сохраняем результаты для бэка

In [23]:
preds.to_dataframe().to_csv('predicts/predict_v01.csv', index=False )

### Делаем предсказание для Test

### Исключаем предсказания групп, которых не будет в марте-апреле

### Сохраняем test.csv